pkg/generators/forwarding/fluentd/generators.go

```diff
@@ -161,11 +161,27 @@ func gatherSources(forwarder *logging.ClusterLogForwarderSpec) (types sets.Strin
 	return types, namespaces
 }
 
-func inputsToPipelines(pipelines []logging.PipelineSpec) logging.RouteMap {
+func inputsToPipelines(fwdspec *logging.ClusterLogForwarderSpec) logging.RouteMap {
 	result := logging.RouteMap{}
-	for _, pipeline := range pipelines {
+	for _, pipeline := range fwdspec.Pipelines {
 		for _, inRef := range pipeline.InputRefs {
```

Comment:
This is correct, but I propose the following refactor. Getting the types from an InputSpec seem like it might be a common operation so I put it into the logging package. This simplifies the generator code which can also handle more than one type in an InputSpec (not sure if we will ever need that but it's easy so might as well play safe.)
```
diff --git a/pkg/apis/logging/v1/cluster_log_forwarder.go b/pkg/apis/logging/v1/cluster_log_forwarder.go
index ec98429f..dbd84db9 100644
--- a/pkg/apis/logging/v1/cluster_log_forwarder.go
+++ b/pkg/apis/logging/v1/cluster_log_forwarder.go
@@ -108,3 +108,18 @@ func (spec *ClusterLogForwarderSpec) HasDefaultOutput() bool {
 	_, ok := spec.OutputMap()[OutputNameDefault]
 	return ok
 }
+
+// Types returns the set of input types that are used to by the input spec.
+func (input *InputSpec) Types() sets.String {
+	result := sets.NewString()
+	if input.Application != nil {
+		result.Insert(InputNameApplication)
+	}
+	if input.Infrastructure != nil {
+		result.Insert(InputNameInfrastructure)
+	}
+	if input.Audit != nil {
+		result.Insert(InputNameAudit)
+	}
+	return result
+}
diff --git a/pkg/generators/forwarding/fluentd/generators.go b/pkg/generators/forwarding/fluentd/generators.go
index c28c0184..42cfd0d3 100644
--- a/pkg/generators/forwarding/fluentd/generators.go
+++ b/pkg/generators/forwarding/fluentd/generators.go
@@ -163,24 +163,15 @@ func gatherSources(forwarder *logging.ClusterLogForwarderSpec) (types sets.Strin
 
 func inputsToPipelines(fwdspec *logging.ClusterLogForwarderSpec) logging.RouteMap {
 	result := logging.RouteMap{}
+	inputs := fwdspec.InputMap()
 	for _, pipeline := range fwdspec.Pipelines {
 		for _, inRef := range pipeline.InputRefs {
-			if len(fwdspec.Inputs) > 0 {
-				// look up type of input
-				for _, input := range fwdspec.Inputs {
-					if input.Name == inRef {
-						// ideally input should only be of one type
-						if input.Application != nil {
-							result.Insert(logging.InputNameApplication, pipeline.Name)
-						} else if input.Infrastructure != nil {
-							result.Insert(logging.InputNameInfrastructure, pipeline.Name)
-						} else if input.Audit != nil {
-							result.Insert(logging.InputNameAudit, pipeline.Name)
-						}
-					}
+			if input, ok := inputs[inRef]; ok { // User defined input spec, unwrap.
+				for t := range input.Types() {
+					result.Insert(t, pipeline.Name)
 				}
 			} else {
-				result.Insert(inRef, pipeline.Name)
+				result.Insert(inRef, pipeline.Name) // Not a user defined type, insert direct.
 			}
 		}
 	}
```

Reply:
updated.